include/nccl_ofi_ofiutils.h

```diff
@@ -8,30 +8,77 @@
 #include <rdma/fabric.h>
 
 #include "nccl_ofi_param.h"
+#include "ofi/nccl_ofi_ofiwrapper.h"
 
 int nccl_ofi_ofiutils_get_providers(const char *prov_include,
 				    uint32_t required_version,
 				    struct fi_info *hints,
 				    struct fi_info **prov_info_list,
 				    unsigned int *num_prov_infos);
 
+/**
+ * @brief	Release libfabric endpoint and address vector
+ */
+void nccl_ofi_ofiutils_ep_release(fid_ep_ptr& ep, fid_av_ptr& av, int dev_id);
 
-/*
- * @brief	Allocates and initialises libfabric endpoint and AV.
+/**
+ * @brief	Create and initialize libfabric fabric
  *
- * @param cq:	Completion queue to which the new endpoint will be bound
- * @return	Endpoint ep
- * @return	Address vector av
+ * @param info:		Fabric info for fabric creation
+ * @return		Smart pointer for fabric on success, nullptr on failure
  */
-int nccl_ofi_ofiutils_init_connection(struct fi_info *info, struct fid_domain *domain,
-				      struct fid_ep **ep,   struct fid_av **av,
-				      struct fid_cq *cq);
+fid_fabric_ptr nccl_ofi_ofiutils_fabric_create(struct fi_info *info);
 
-/*
- * @brief	Release libfabric endpoint and address vector
+/**
+ * @brief	Create and initialize libfabric domain
+ *
+ * @param fabric:	Fabric handle
+ * @param info:		Fabric info for domain creation
+ * @return		Smart pointer for domain on success, nullptr on failure
+ */
+fid_domain_ptr nccl_ofi_ofiutils_domain_create(struct fid_fabric *fabric, struct fi_info *info);
+
+/**
+ * @brief	Create and initialize libfabric endpoint
+ *
+ * @param info:		Fabric info for endpoint creation
+ * @param domain:	Fabric domain
+ * @param av:		Address vector to which the new endpoint will be bound
+ * @param cq:		Completion queue to which the new endpoint will be bound
+ * @return		Smart pointer for endpoint on success, nullptr on failure
+ */
+fid_ep_ptr nccl_ofi_ofiutils_ep_create(struct fi_info *info, struct fid_domain *domain,
+				       struct fid_av *av, struct fid_cq *cq);
+
+/**
+ * @brief	Create and initialize libfabric address vector
+ *
+ * @param domain:	Domain handle
+ * @return		Smart pointer for address vector on success, nullptr on failure
+ */
+fid_av_ptr nccl_ofi_ofiutils_av_create(struct fid_domain *domain);
+
+/**
+ * @brief	Create and initialize libfabric completion queue
+ *
+ * @param domain:	Domain handle
+ * @param cq_size:	Size of completion queue
+ * @param info:		Fabric info for cq creation
+ * @return		Smart pointer for completion queue on success, nullptr on failure
+ */
+fid_cq_ptr nccl_ofi_ofiutils_cq_create(struct fid_domain *domain, struct fi_cq_attr *cq_attr);
```

Comment:
If we're going to create these wrappers, shouldn't they take the first argument as the unique_ptr by reference?